gobblin-runtime/src/main/java/gobblin/runtime/JobContext.java

```diff
@@ -350,77 +351,105 @@ void commit() throws IOException {
+          datasetState.setState(JobState.RunningState.COMMITTED);
         } else {
-          if (datasetState.getState() == JobState.RunningState.SUCCESSFUL) {
-            datasetState.setState(JobState.RunningState.COMMITTED);
-          }
+          persistDatasetState(datasetUrn, datasetState);
         }
-      } catch (ReflectiveOperationException roe) {
+      } catch (IOException|RuntimeException ioe) {
         this.logger.error(
-            String.format("Failed to instantiate data publisher for dataset %s of job %s.", datasetUrn, this.jobId),
-            roe);
-      } catch (IOException ioe) {
-        this.logger.error(
-            String.format("Failed to commit dataset state for dataset %s of job %s", datasetUrn, this.jobId), ioe);
-        allDatasetsCommit = false;
-      } finally {
-        try {
-          finalizeDatasetState(datasetState, datasetUrn);
-
-          if (this.commitSequenceBuilder.isPresent()) {
-            buildAndExecuteCommitSequence(this.commitSequenceBuilder.get(), datasetState, datasetUrn);
-            datasetState.setState(JobState.RunningState.COMMITTED);
-          } else {
-            persistDatasetState(datasetUrn, datasetState);
-          }
-        } catch (IOException ioe) {
-          this.logger.error(
-              String.format("Failed to persist dataset state for dataset %s of job %s", datasetUrn, this.jobId), ioe);
-          allDatasetsCommit = false;
-        }
+            String.format("Failed to persist dataset state for dataset %s of job %s", datasetUrn, this.jobId), ioe);
+        success = false;
       }
     }
+    return success;
+  }
 
-    if (!allDatasetsCommit) {
-      this.jobState.setState(JobState.RunningState.FAILED);
-      throw new IOException("Failed to commit dataset state for some dataset(s) of job " + this.jobId);
+  boolean checkForUnpublishedWUHandling(String datasetUrn, JobState.DatasetState datasetState,
+                                        Class<? extends DataPublisher> dataPublisherClass,
+                                        Closer closer)
+          throws ReflectiveOperationException, IOException {
+    boolean success = true;
+    if (!canCommitDataset(datasetState)) {
+      this.logger.warn(String.format("Not committing dataset %s of job %s with commit policy %s and state %s",
+          datasetUrn, this.jobId, this.jobCommitPolicy, datasetState.getState()));
+      if (UnpublishedHandling.class.isAssignableFrom(dataPublisherClass)) {
+        DataPublisher publisher = closer.register(DataPublisher.getInstance(dataPublisherClass, datasetState));
+        this.logger
+            .info(String.format("Calling publisher to handle unpublished work units for dataset %s of job %s.",
+                datasetUrn, this.jobId));
+        ((UnpublishedHandling) publisher).handleUnpublishedWorkUnits(datasetState.getTaskStatesAsWorkUnitStates());
+      }
+      success = false;
```

Comment:
Preserves existing logic. I'll choose better naming. The return result is if the job can be committed.